tensorflow/lite/micro/docs/rfc/003_mem_alloc_updates.md

```diff
@@ -0,0 +1,52 @@
+|               | Niranjan Yadla (nyadla@cadence.com)                         |
+|               | Raj Pawate (pawateb@cadence.com)                            |
+|               | Vijay Pawar (vpawar@cadence.com)                            |
+| **Sponsor**   | Pete Warden (petewarden@google.com)                         |
+| **Updated**   | 2021-05-12                                                  |
+
+## Background
+
+TensorFlow Lite for Microcotrollers (TFLM) makes deploying of neural networks on DSP very easy. For edge devices, DSP would generally run multiple neural networks simultaneously and would also run DSP workload for frontend decode of neural networks, music playback, speech encode - decode etc. To enable quick and efficient deployment of these mix workloads on DSP, some framework would be useful. We are planning to use Cadence's Xtensa Audio Framework (XAF) for this purpose.
+The current memory allocation scheme in TFLM is not favorable for integration into XAF, so following updates are being proposed.
+
+## Current TFLM Memory Allocations
+In current TFLM implementation, one large tensor arena buffer is passed to MicroInterpreter. MicroInterpreter then allocates various tensors and buffers required for the network from this arena through init and prepare calls of operators. Scratch allocations are done from the head of the arena and persistent allocations are done from the tail of the arena.
+Note, scratch memory is used by the network during one invoke call and can be reused outside of the invoke call whereas persistent memory is used by the network during its lifetime and cannot be reused for any other purpose.
+While, the current allocation scheme is clean and easy to use, following issues can be observed:
+- Input and output tensors of the network are within arena. Application / upper layer would need to copy data to and from these buffers on each invoke call.
+- It may not be possible to reuse scratch memory for other DSP workload.
+- The upfront memory requirement of network is not known.
+
+## Proposed Updates in TFLM Memory Allocations
+Following updates in TFLM memory allocation scheme would really make it efficient and more attractive for embedded frameworks.
+- Provide a query API to get input, output, scratch and persistent memory requirements of the network at init time.
+- Allow scratch and persistent buffers to be separate and respective buffer pointers would be passed to MicroInterpreter.
+- Allow input and output buffers to be allocated by application / upper layer and respective buffer pointers would be passed to MicroInterpreter.
+
+A query API can be something like below:
+```
+tflite::void GetMemRequirements(const Model* model, 
+                                int32_t* inp_sz,
+                                int32_t* out_sz,
+                                int32_t* persist_sz,
+                                int32_t* scratch_sz);
```

Comment:
Is allocating a single contiguous buffer for each of the scratch/persistent/in/out areas granular enough across all the platforms we might need to support? In particular I'm thinking of DSPs that have X/Y memory and so need to allocate activation buffers in alternating memory banks to get optimal performance.

As a very brief sketch of an alternative proposal, would your needs be satisfied if:

  - Input/Output allocation areas could be overridden by the application at runtime. The client already has to know the sizes of these, so I don't think a query mechanism would be needed.
  - The location of activation buffers could be overridden by specifying them in the flatbuffer as metadata. This would be possible offline because the sizes are known at that point, so no query should be needed.
  - Persistent and op-requested scratch buffers are still allocated from the arena, since their requirements aren't known until runtime.

In particular, I'd like to know if "scratch buffers" (in my terms those allocated by optimized ops, and so unknown in size offline) or "activation buffers" (the temporary input/output intermediates of ops, which are of a known size) are the biggest issue for you?

Reply:
Yes, allocating a single contiguous buffer for each of the scratch/persistent/in/out areas should suffice for XAF integration. Most Cadence DSPs have 2 load/store units and having X/Y memory granularity would be helpful, but as coefficients / filters come from different memory (flatbuffer) in TFLM, some X/Y memory optimizations can be done anyway.

Thanks for the suggestions and I think overriding input/ output pointers at runtime would work.
A query API would still be very useful as it makes code portable and avoids offline model adjustments or code modifications anytime something changes in operator or network.
As you pointed out, some of the persist and op-requested scratch buffers are known only at runtime.
If supported, a dry run through init and prepare calls of operators in the network can report all the memory requirements.

Between "scratch buffers" and "activation buffers", putting activation buffers in separate scratch memory would be more important as I reckon that would be a bigger memory chunk.